File: java/src/test/java/de/train/bewerbung/JdbcBewerbungRepositoryTest.java
package de.train.bewerbung;

import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;
import java.sql.Statement;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * INTEGRATION-Test: laeuft gegen die ECHTE MariaDB.
 * Jeder Test in einer Transaktion, die am Ende zurueckgerollt
 * wird -> die Datenbank bleibt unveraendert.
 */
class JdbcBewerbungRepositoryTest {

    private Connection conn;
    private JdbcBewerbungRepository repo;

    private static Connection connect() throws SQLException {
        String host = System.getenv("DB_HOST");
        String name = System.getenv("DB_NAME");
        String user = System.getenv("DB_USER");
        String pass = System.getenv("DB_PASS");
        String url = "jdbc:mariadb://" + host + ":3306/" + name
                + "?useUnicode=true&characterEncoding=utf8";
        return DriverManager.getConnection(url, user, pass);
    }

    @BeforeEach
    void setUp() throws SQLException {
        conn = connect();
        conn.setAutoCommit(false);
        repo = new JdbcBewerbungRepository(conn);
    }

    @AfterEach
    void tearDown() throws SQLException {
        if (conn != null) {
            conn.rollback();
            conn.close();
        }
    }

    private int eineStelleId() throws SQLException {
        try (Statement st = conn.createStatement()) {
            st.executeUpdate(
                    "INSERT INTO stellenangebot (titel, art, status) "
                            + "VALUES ('Test-Stelle', 'FESTANSTELLUNG', 'VEROEFFENTLICHT')",
                    Statement.RETURN_GENERATED_KEYS);
            try (var keys = st.getGeneratedKeys()) {
                keys.next();
                return keys.getInt(1);
            }
        }
    }

    @Test
    void bewerberAnlegenUndPerEmailFinden() throws SQLException {
        int id = repo.insertBewerber(new BewerberInput(
                "Erika", "Mustermann", "java-int@example.com", null));
        assertEquals(Integer.valueOf(id),
                repo.findBewerberIdByEmail("java-int@example.com"));
        assertNull(repo.findBewerberIdByEmail("unbekannt@example.com"));
    }

    @Test
    void bewerbungAnlegenFunktioniert() throws SQLException {
        int stelleId = eineStelleId();
        int bid = repo.insertBewerber(new BewerberInput(
                "Max", "M", "java-m@example.com", null));
        int id = repo.insertBewerbung(bid, stelleId, "BEW-2026-JAVA01", null);
        assertTrue(id > 0);
    }

    @Test
    void fremdschluesselVerhindertUngueltigeStelle() throws SQLException {
        int bid = repo.insertBewerber(new BewerberInput(
                "A", "B", "java-fk@example.com", null));
        SQLException ex = assertThrows(SQLException.class,
                () -> repo.insertBewerbung(bid, 999999, "BEW-2026-JAVAFK", null));
        assertEquals(1452, ex.getErrorCode());
    }

    @Test
    void vorgangsnummerIstEindeutig() throws SQLException {
        int stelleId = eineStelleId();
        int bid = repo.insertBewerber(new BewerberInput(
                "C", "D", "java-uq@example.com", null));
        repo.insertBewerbung(bid, stelleId, "BEW-2026-JADUP1", null);
        SQLException ex = assertThrows(SQLException.class,
                () -> repo.insertBewerbung(bid, stelleId, "BEW-2026-JADUP1", null));
        assertEquals(1062, ex.getErrorCode());
    }
}

File: kotlin/src/test/kotlin/de/train/bewerbung/JdbcBewerbungRepositoryTest.kt
package de.train.bewerbung

import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.sql.Connection
import java.sql.SQLException
import java.sql.Statement

/**
 * INTEGRATION-Test gegen die ECHTE MariaDB.
 * Jeder Test in einer Transaktion, die zurueckgerollt wird.
 */
class JdbcBewerbungRepositoryTest {

    private lateinit var conn: Connection
    private lateinit var repo: JdbcBewerbungRepository

    @BeforeEach
    fun setUp() {
        conn = openConnection()
        conn.autoCommit = false
        repo = JdbcBewerbungRepository(conn)
    }

    @AfterEach
    fun tearDown() {
        conn.rollback()
        conn.close()
    }

    private fun eineStelleId(): Int {
        conn.createStatement().use { st ->
            st.executeUpdate(
                "INSERT INTO stellenangebot (titel, art, status) " +
                        "VALUES ('Test-Stelle', 'FESTANSTELLUNG', 'VEROEFFENTLICHT')",
                Statement.RETURN_GENERATED_KEYS
            )
            st.generatedKeys.use { rs -> rs.next(); return rs.getInt(1) }
        }
    }

    @Test
    fun `bewerber anlegen und per email finden`() {
        val id = repo.insertBewerber(BewerberInput(
            "Erika", "Mustermann", "kt-int@example.com", null))
        assertEquals(id, repo.findBewerberIdByEmail("kt-int@example.com"))
        assertNull(repo.findBewerberIdByEmail("unbekannt@example.com"))
    }

    @Test
    fun `bewerbung anlegen funktioniert`() {
        val stelleId = eineStelleId()
        val bid = repo.insertBewerber(BewerberInput("Max", "M", "kt-m@example.com", null))
        val id = repo.insertBewerbung(bid, stelleId, "BEW-2026-KTLN01", null)
        assertTrue(id > 0)
    }

    @Test
    fun `fk verhindert ungueltige stelle`() {
        val bid = repo.insertBewerber(BewerberInput("A", "B", "kt-fk@example.com", null))
        val ex = assertThrows(SQLException::class.java) {
            repo.insertBewerbung(bid, 999999, "BEW-2026-KTFK01", null)
        }
        assertEquals(1452, ex.errorCode)
    }

    @Test
    fun `vorgangsnummer ist eindeutig`() {
        val stelleId = eineStelleId()
        val bid = repo.insertBewerber(BewerberInput("C", "D", "kt-uq@example.com", null))
        repo.insertBewerbung(bid, stelleId, "BEW-2026-KTDUP1", null)
        val ex = assertThrows(SQLException::class.java) {
            repo.insertBewerbung(bid, stelleId, "BEW-2026-KTDUP1", null)
        }
        assertEquals(1062, ex.errorCode)
    }
}
